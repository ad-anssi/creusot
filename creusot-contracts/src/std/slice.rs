use crate as creusot_contracts;
use creusot_contracts_proc::*;

use crate::OrdLogic;

extern_spec! {
  #[ensures((@s).len() === @result)]
  fn <[T]>::len<T>(s: &[T]) -> usize
}

extern_spec! {
    #[requires(@i < (@s).len())]
    #[requires(@j < (@s).len())]
    #[ensures((@^s).exchange(@*s, @i, @j))]
    fn <[T]>::swap<T>(s: &mut [T], i: usize, j: usize)
}