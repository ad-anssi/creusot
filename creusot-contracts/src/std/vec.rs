use crate as creusot_contracts;
use crate::logic::*;
use crate::{std::clone::Clone, Int, Model, Seq};
use creusot_contracts_proc::*;

impl<T> Model for Vec<T> {
    type ModelTy = Seq<T>;
    #[logic]
    #[trusted]
    fn model(self) -> Self::ModelTy {
        std::process::abort()
    }
}

extern_spec! {
    #[ensures((@result).len() === 0)]
    fn std::vec::Vec::new<T>() -> std::vec::Vec<T>
}

extern_spec! {
    fn std::vec::Vec::len<T>(v: &std::vec::Vec<T>) -> usize
}

extern_spec! {
    #[ensures(result.into() === (@v).len())]
    fn std::vec::Vec::len<T>(v: &Vec<T>) -> usize
}


extern_spec! {
    #[ensures(@^self_ === (@self_).push(v))]
    fn std::vec::Vec::push<T>(self_: &mut Vec<T>, v: T)
}

// extern_spec! {
//     #[requires(@i < (@self_).len())]
//     #[requires(@j < (@self_).len())]
//     #[ensures((@^self_).exchange(@*self_, @i, @j))]
//     fn std::vec::Vec::swap<T>(self_: &mut Vec<T>, i: usize, j: usize)
// }

// impl<T> Vec<T> {
//     #[trusted]
//     #[ensures((@result).len() === 0)]
//     pub fn new() -> Self {
//         Vec(std::vec::Vec::new())
//     }

//     #[trusted]
//     #[ensures((@result).len() === 0)]
//     pub fn with_capacity(capacity: usize) -> Vec<T> {
//         Vec(std::vec::Vec::with_capacity(capacity))
//     }

//     #[trusted]
//     #[ensures(result.into() === (@self).len())]
//     pub fn len(&self) -> usize {
//         self.0.len()
//     }

//     #[trusted]
//     #[ensures(match result {
//         Some(t) => *t === (@*self)[ix.into()],
//         None => (@*self).len() <= ix.into(),
//     })]
//     pub fn get(&self, ix: usize) -> Option<&T> {
//         self.0.get(ix)
//     }

//     #[trusted]
//     #[ensures(@^self === (@self).push(v))]
//     pub fn push(&mut self, v: T) {
//         self.0.push(v)
//     }

//     #[trusted]
//     #[requires(@i < (@self).len())]
//     #[requires(@j < (@self).len())]
//     #[ensures((@^self).exchange(@*self, @i, @j))]
//     pub fn swap(&mut self, i: usize, j: usize) {
//         self.0.swap(i, j)
//     }

//     #[trusted]
//     #[ensures(match result {
//         Some(t) => (@self) === (@^self).push(t),
//         None => (@self).len() === (@^self).len() && (@self).len() === 0
//     })]
//     pub fn pop(&mut self) -> Option<T> {
//         self.0.pop()
//     }
// }

// impl<T> std::ops::IndexMut<usize> for Vec<T> {
//     #[trusted]
//     #[requires(@ix < (@*self).len())]
//     #[ensures(*result === (@self)[@ix])]
//     #[ensures(^result === (@^self)[@ix])]
//     #[ensures(forall<j : Int> 0 <= j && j < (@^self).len() ==>
//         !(j === @ix) ==>
//         (@^self)[j] === (@*self)[j])]
//     #[ensures((@*self).len() === (@^self).len())]
//     fn index_mut(&mut self, ix: usize) -> &mut T {
//         self.0.index_mut(ix)
//     }
// }

use std::slice::SliceIndex;
use std::ops::{Index, IndexMut};
// impl<T, I> std::ops::Index<I> for Vec<T>
//     where I : SliceIndex<[T]> {
//     type Output = <I as SliceIndex<[T]>>::Output;

//     #[trusted]
//     // #[requires(@ix < (@self).len())]
//     // #[ensures(*result === (@self)[@ix])]
//     fn index(&self, ix: I) -> &T {
//         self.0.index(ix)
//     }
// }
//
trait IndexSpec<I> : Index<I> {

    // Check whether an index is 'in bounds' for a structure
    #[predicate]
    fn in_bounds(self, i: I) -> bool;

    // Condition underwhich we get `out` from index `i` in `self`
    #[predicate]
    fn read(self, i: I, out: Self::Output) -> bool;
}

extern_spec! {
    #[requires(self_.in_bounds(i))]
    #[ensures(self_.read(i, *result))]
    fn std::ops::Index::index<T, I>(self_: &T, i: I) -> &T::Output
        where T : IndexSpec<I>
}

trait IndexMutSpec<I> : IndexMut<I> {
    // Check whether an index is 'in bounds' for a structure
    #[predicate]
    fn in_bounds(self, i: I) -> bool;

    // Condition underwhich we get `out` from index `i` in `self`
    #[predicate]
    fn read(self, i: I, out: Self::Output) -> bool;

    // Explains what happens to the elements we didn't index
    #[predicate]
    fn context(&mut self, i: I) -> bool;
}


extern_spec! {
    #[requires(self_.in_bounds(i))]
    #[ensures((*self_).read(i, *result))]
    #[ensures((^self_).read(i, ^result))]
    #[ensures(self_.context(i))]
    fn std::ops::IndexMut::index_mut<T, I>(self_: &mut T, i: I) -> &mut T::Output
        where
                T : IndexMutSpec<I>,
                T : IndexMut<I>,
}

trait SeqIndex<T> : SliceIndex<[T]> {
    // Check whether an index is 'in bounds' for a structure
    #[predicate]
    fn in_bounds(self, s: Seq<T>) -> bool;

    // Condition underwhich we get `out` from index `i` in `self`
    #[predicate]
    fn read(self, s: Seq<T>, out: Self::Output) -> bool;

    // Explains what happens to the elements we didn't index
    #[predicate]
    fn context(self, old: Seq<T>, new: Seq<T>) -> bool;
}

// We probably want to move these into a `SliceIndexSpec` trait as well...
impl<T, I : SeqIndex<T>> IndexMutSpec<I> for Vec<T> {
    #[predicate]
    fn in_bounds(self, i : I) -> bool {
        pearlite! { i.in_bounds(@self) }
    }

    #[predicate]
    fn read(self, i: I, out: Self::Output) -> bool {
        pearlite! { i.read(@self, out) }
    }

    #[predicate]
    fn context(&mut self, i : I) -> bool {
        pearlite! { i.context(@*self, @^self) }
    }
}

impl<T> SeqIndex<T> for usize {
    #[predicate]
    fn in_bounds(self, s: Seq<T>) -> bool {
        pearlite! { @self < s.len() }
    }

    #[predicate]
    fn read(self, s: Seq<T>, out: Self::Output) -> bool {
        pearlite! { s[@self] === out }
    }

    #[predicate]
    fn context(self, old: Seq<T>, new: Seq<T>) -> bool {
        true
    }
}


extern_spec! {
    #[requires(self_.in_bounds(ix))]
    #[ensures((*self_).read(ix, *result))]
    #[ensures((^self_).read(ix, ^result))]
    #[ensures(self_.context(ix))]
    fn std::vec::Vec::index_mut<T, I : SeqIndex<T>>(self_: &mut Vec<T>, ix: I) -> &mut <I as SliceIndex<[T]>>::Output
        where T : IndexMutSpec<I>
}

// TODO: Ensure extern functions & extern_specs inherit trait contracts


// extern_spec! {
//   fn std::vec::Vec::index<T, I : SliceIndex<[T]>>(self_: &Vec<T>, ix: I) -> &<I as SliceIndex<[T]>>::Output
//     where crate::Seq<T> : Index<I, Output = <I as SliceIndex<[T]>>::Output>
// }

// trait SliceIndexSpec<T> : SliceIndex<[T]> {
//     // type Output;

//     #[predicate]
//     fn in_bounds(self, seq: Seq<T>) -> bool;

//     // TODO: better name
//     #[predicate]
//     fn partition(self, seq: Seq<T>, res: Self::Output) -> bool;
// }

// impl<T> SliceIndexSpec<T> for usize {
//     #[predicate]
//     fn in_bounds(self, seq: Seq<T>) -> bool {
//         pearlite! { @self < seq.len() }
//     }

//     #[predicate]
//     fn partition(self, seq: Seq<T>, res: Self::Output) -> bool {
//         pearlite! { seq.set(@self, res) }
//     }
// }


// extern_spec! {
//   fn std::vec::Vec::index<T, I : SliceIndex<[T]>>(self_: &Vec<T>, ix: I) -> &<I as SliceIndex<[T]>>::Output
//     where crate::Seq<T> : Index<I, Output = <I as SliceIndex<[T]>>::Output>
// }

// extern_spec! {
//   #[requires(ix.in_bounds(@self_))]
//   #[ensures(ix.partition(@*self_, *result))]
//   #[ensures(ix.partition(@^self_, ^result))]
//   fn std::vec::Vec::index_mut<T, I : SliceIndexSpec<T>>(self_: &mut Vec<T>, ix: I) -> &mut <I as SliceIndex<[T]>>::Output
//     // where crate::Seq<T> : Index<I>,
//     //       <I as SliceIndex<[T]>>::Output : Model,
// }

// impl<T: Clone> Clone for Vec<T> {
//     #[trusted]
//     fn clone(&self) -> Self {
//         panic!()
//         // Vec(self.0.iter().map(|r : &T| r.clone()).collect())
//     }
// }

unsafe impl<T> Resolve for Vec<T> {
    #[predicate]
    fn resolve(self) -> bool {
        pearlite! { forall<i : Int> 0 <= i && i < (@self).len() ==> (@self)[i].resolve() }
    }
}

// #[trusted]
// #[ensures((@result).len() === @n)]
// #[ensures(forall<i : Int> 0 <= i && i < @n ==> (@result)[i] === elem)]
// pub fn from_elem<T: Clone>(elem: T, n: usize) -> Vec<T> {
//     panic!()
// }
