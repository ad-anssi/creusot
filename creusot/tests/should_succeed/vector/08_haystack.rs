extern crate creusot_contracts;
use creusot_contracts::*;
use creusot_contracts::std::*;

#[predicate]
fn match_at(needle: &Vec<u8>, haystack: &Vec<u8>, pos: Int, len: Int) -> bool {
  pearlite! { len <= (@needle).len()
    && pos <= (@haystack).len() - len
    && forall<i: Int>
        0 <= i && i < len ==> (@needle)[i] === (@haystack)[pos + i]
  }
}

#[requires((@needle).len() >= 1 && (@needle).len() <= (@haystack).len())]
#[ensures(@result == (@haystack).len() || @result < (@haystack).len() - (@needle).len() + 1)]
#[ensures(@result < (@haystack).len() ==>
            (match_at(needle, haystack, @result, (@needle).len())
             && (forall <i: Int> 0 <= i && i < @result ==> ! match_at(needle, haystack, i, (@needle).len()))))]
#[ensures(@result == (@haystack).len() ==> (forall <i: Int> 0 <= i && i < (@haystack).len() ==> ! match_at(needle, haystack, i, (@needle).len())))]
fn search(needle: &Vec<u8>, haystack: &Vec<u8>) -> usize {
   let mut i:usize = 0;
   // #[invariant(range_i,@i <= (@haystack).len() - (@needle).len() + 1)]
   #[invariant(no_match,forall<k: Int> 0 <= k && k < @i ==> ! match_at(needle, haystack, k, (@needle).len()))]
   while i < haystack.len() - needle.len() + 1 {
      let mut j:usize = 0;
      // #[invariant(range_j,@j <= (@needle).len())]
      #[invariant(partial_match,match_at(needle, haystack, @i, @j))]
      while j < needle.len() {
         proof_assert!(@j <= (@needle).len());
         if needle[j] != haystack[i + j] {
            break
         }
         j += 1;
      };
      proof_assert! { false};
      if j == needle.len() {
         return i
      };

      i += 1
   };
   return haystack.len()
}
