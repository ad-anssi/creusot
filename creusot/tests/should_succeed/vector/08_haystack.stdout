module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type creusotcontracts_std1_vec_vec 't  
end
module CreusotContracts_Logic_Model_Model_ModelTy
  type self   
  type modelTy   
end
module CreusotContracts_Logic_Model_Model_Model_Interface
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Model_Model
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl0_Model_Interface
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  function model (self : t) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl0_Model
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  clone CreusotContracts_Logic_Model_Model_Model_Interface as Model0 with type self = t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  function model (self : t) : ModelTy0.modelTy = 
    Model0.model self
end
module CreusotContracts_Std1_Vec_Impl0_ModelTy
  type t   
  use seq.Seq
  type modelTy  = 
    Seq.seq t
end
module C08Haystack_MatchAt_Interface
  use prelude.Prelude
  use Type
  use mach.int.Int
  use prelude.UInt8
  predicate match_at (needle : Type.creusotcontracts_std1_vec_vec uint8) (haystack : Type.creusotcontracts_std1_vec_vec uint8) (pos : int) (len : int)
    
end
module C08Haystack_MatchAt
  use prelude.Prelude
  use Type
  use mach.int.Int
  use prelude.UInt8
  use seq.Seq
  use mach.int.Int32
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = uint8
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec uint8,
  type ModelTy0.modelTy = ModelTy0.modelTy
  predicate match_at (needle : Type.creusotcontracts_std1_vec_vec uint8) (haystack : Type.creusotcontracts_std1_vec_vec uint8) (pos : int) (len : int)
    
   = 
    len <= Seq.length (Model0.model needle) && pos <= Seq.length (Model0.model haystack) - len && (forall i : (int) . 0 <= i && i < len -> Seq.get (Model0.model needle) i = Seq.get (Model0.model haystack) (pos + i))
end
module CreusotContracts_Logic_Model_Impl0_ModelTy
  type t   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  type modelTy  = 
    ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl0
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy2 with type self = t
  clone CreusotContracts_Logic_Model_Model_Model as Model2 with type self = t, type ModelTy0.modelTy = ModelTy2.modelTy
  clone CreusotContracts_Logic_Model_Impl0_Model as Model0 with type t = t, type ModelTy0.modelTy = ModelTy2.modelTy,
  function Model0.model = Model2.model
  clone CreusotContracts_Logic_Model_Impl0_ModelTy as ModelTy0 with type t = t, type ModelTy0.modelTy = ModelTy2.modelTy
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = t, function model = Model0.model,
  type ModelTy0.modelTy = ModelTy0.modelTy
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = t, type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Std1_Vec_Impl0_Model_Interface
  type t   
  use Type
  use seq.Seq
  function model (self : Type.creusotcontracts_std1_vec_vec t) : Seq.seq t
end
module CreusotContracts_Std1_Vec_Impl0_Model
  type t   
  use Type
  use seq.Seq
  function model (self : Type.creusotcontracts_std1_vec_vec t) : Seq.seq t
end
module CreusotContracts_Std1_Vec_Impl0
  type t   
  use Type
  clone CreusotContracts_Std1_Vec_Impl0_Model as Model0 with type t = t
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = Type.creusotcontracts_std1_vec_vec t,
  function model = Model0.model, type ModelTy0.modelTy = ModelTy0.modelTy
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = Type.creusotcontracts_std1_vec_vec t,
  type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve_Interface
  type t   
  predicate resolve (self : t)
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve
  type t   
  predicate resolve (self : t) = 
    true
end
module CreusotContracts_Std1_Vec_Impl1_Len_Interface
  type t   
  use mach.int.UInt64
  use seq.Seq
  use prelude.Prelude
  use Type
  use mach.int.Int
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val len [@cfg:stackify] (self : Type.creusotcontracts_std1_vec_vec t) : usize
    ensures { UInt64.to_int result = Seq.length (Model0.model self) }
    
end
module CreusotContracts_Std1_Vec_Impl1_Len
  type t   
  use mach.int.UInt64
  use seq.Seq
  use prelude.Prelude
  use Type
  use mach.int.Int
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val len [@cfg:stackify] (self : Type.creusotcontracts_std1_vec_vec t) : usize
    ensures { UInt64.to_int result = Seq.length (Model0.model self) }
    
end
module Core_Ops_Index_Index_Output
  type self   
  type idx   
  type output   
end
module Core_Ops_Index_Index_Index_Interface
  type self   
  type idx   
  use prelude.Prelude
  clone Core_Ops_Index_Index_Output as Output0 with type self = self, type idx = idx
  val index [@cfg:stackify] (self : self) (index : idx) : Output0.output
    requires {false}
    
end
module Core_Ops_Index_Index_Index
  type self   
  type idx   
  use prelude.Prelude
  clone Core_Ops_Index_Index_Output as Output0 with type self = self, type idx = idx
  val index [@cfg:stackify] (self : self) (index : idx) : Output0.output
    requires {false}
    
end
module CreusotContracts_Std1_Vec_Impl3_Index_Interface
  type t   
  use mach.int.UInt64
  use seq.Seq
  use mach.int.Int
  use prelude.Prelude
  use Type
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val index [@cfg:stackify] (self : Type.creusotcontracts_std1_vec_vec t) (ix : usize) : t
    requires {UInt64.to_int ix < Seq.length (Model0.model self)}
    ensures { result = Seq.get (Model0.model self) (UInt64.to_int ix) }
    
end
module CreusotContracts_Std1_Vec_Impl3_Index
  type t   
  use mach.int.UInt64
  use seq.Seq
  use mach.int.Int
  use prelude.Prelude
  use Type
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val index [@cfg:stackify] (self : Type.creusotcontracts_std1_vec_vec t) (ix : usize) : t
    requires {UInt64.to_int ix < Seq.length (Model0.model self)}
    ensures { result = Seq.get (Model0.model self) (UInt64.to_int ix) }
    
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Impl2
  type t   
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve0 with type t = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = t,
  predicate resolve = Resolve0.resolve
end
module CreusotContracts_Std1_Vec_Impl3_Output
  type t   
  type output  = 
    t
end
module CreusotContracts_Std1_Vec_Impl3
  type t   
  use Type
  use mach.int.Int
  use prelude.Prelude
  use mach.int.UInt64
  clone CreusotContracts_Std1_Vec_Impl0_Model as Model1 with type t = t
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy, function Model0.model = Model1.model
  clone CreusotContracts_Std1_Vec_Impl3_Index_Interface as Index0 with type t = t, function Model0.model = Model0.model
  clone CreusotContracts_Std1_Vec_Impl3_Output as Output0 with type t = t
  clone Core_Ops_Index_Index_Index_Interface as Index1 with type self = Type.creusotcontracts_std1_vec_vec t,
  type idx = usize, val index = Index0.index, type Output0.output = Output0.output
  clone Core_Ops_Index_Index_Output as Output1 with type self = Type.creusotcontracts_std1_vec_vec t, type idx = usize,
  type output = Output0.output
end
module C08Haystack_Search_Interface
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use mach.int.UInt64
  use prelude.Prelude
  use Type
  use prelude.UInt8
  clone C08Haystack_MatchAt_Interface as MatchAt0
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = uint8
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec uint8,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val search [@cfg:stackify] (needle : Type.creusotcontracts_std1_vec_vec uint8) (haystack : Type.creusotcontracts_std1_vec_vec uint8) : usize
    requires {Seq.length (Model0.model needle) >= 1 && Seq.length (Model0.model needle) <= Seq.length (Model0.model haystack)}
    ensures { UInt64.to_int result = Seq.length (Model0.model haystack) -> (forall i : (int) . 0 <= i && i < Seq.length (Model0.model haystack) -> not (MatchAt0.match_at needle haystack i (Seq.length (Model0.model needle)))) }
    ensures { UInt64.to_int result < Seq.length (Model0.model haystack) -> MatchAt0.match_at needle haystack (UInt64.to_int result) (Seq.length (Model0.model needle)) && (forall i : (int) . 0 <= i && i < UInt64.to_int result -> not (MatchAt0.match_at needle haystack i (Seq.length (Model0.model needle)))) }
    ensures { UInt64.to_int result = Seq.length (Model0.model haystack) || UInt64.to_int result < Seq.length (Model0.model haystack) - Seq.length (Model0.model needle) + 1 }
    
end
module C08Haystack_Search
  use mach.int.Int
  use mach.int.Int32
  use mach.int.UInt64
  use seq.Seq
  use prelude.Prelude
  use Type
  use prelude.UInt8
  clone CreusotContracts_Std1_Vec_Impl0_Model as Model1 with type t = uint8
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = uint8
  clone CreusotContracts_Logic_Model_Impl0_Model as Model0 with type t = Type.creusotcontracts_std1_vec_vec uint8,
  type ModelTy0.modelTy = ModelTy0.modelTy, function Model0.model = Model1.model
  clone C08Haystack_MatchAt as MatchAt0 with function Model0.model = Model0.model
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve4 with type t = Type.creusotcontracts_std1_vec_vec uint8
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve3 with type t = uint8
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve2 with type t = uint8
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = ()
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve0 with type t = usize
  clone CreusotContracts_Std1_Vec_Impl3_Index_Interface as Index0 with type t = uint8,
  function Model0.model = Model0.model
  clone CreusotContracts_Std1_Vec_Impl1_Len_Interface as Len0 with type t = uint8, function Model0.model = Model0.model
  let rec cfg search [@cfg:stackify] (needle : Type.creusotcontracts_std1_vec_vec uint8) (haystack : Type.creusotcontracts_std1_vec_vec uint8) : usize
    requires {Seq.length (Model0.model needle) >= 1 && Seq.length (Model0.model needle) <= Seq.length (Model0.model haystack)}
    ensures { UInt64.to_int result = Seq.length (Model0.model haystack) -> (forall i : (int) . 0 <= i && i < Seq.length (Model0.model haystack) -> not (MatchAt0.match_at needle haystack i (Seq.length (Model0.model needle)))) }
    ensures { UInt64.to_int result < Seq.length (Model0.model haystack) -> MatchAt0.match_at needle haystack (UInt64.to_int result) (Seq.length (Model0.model needle)) && (forall i : (int) . 0 <= i && i < UInt64.to_int result -> not (MatchAt0.match_at needle haystack i (Seq.length (Model0.model needle)))) }
    ensures { UInt64.to_int result = Seq.length (Model0.model haystack) || UInt64.to_int result < Seq.length (Model0.model haystack) - Seq.length (Model0.model needle) + 1 }
    
   = 
  var _0 : usize;
  var needle_1 : Type.creusotcontracts_std1_vec_vec uint8;
  var haystack_2 : Type.creusotcontracts_std1_vec_vec uint8;
  var i_3 : usize;
  var _4 : ();
  var _5 : ();
  var _6 : ();
  var _7 : bool;
  var _8 : usize;
  var _9 : usize;
  var _10 : usize;
  var _11 : usize;
  var _12 : Type.creusotcontracts_std1_vec_vec uint8;
  var _13 : usize;
  var _14 : Type.creusotcontracts_std1_vec_vec uint8;
  var j_15 : usize;
  var _16 : ();
  var _17 : ();
  var _18 : bool;
  var _19 : usize;
  var _20 : usize;
  var _21 : Type.creusotcontracts_std1_vec_vec uint8;
  var _22 : ();
  var _23 : ();
  var _24 : bool;
  var _25 : uint8;
  var _26 : uint8;
  var _27 : Type.creusotcontracts_std1_vec_vec uint8;
  var _28 : usize;
  var _29 : uint8;
  var _30 : uint8;
  var _31 : Type.creusotcontracts_std1_vec_vec uint8;
  var _32 : usize;
  var _33 : usize;
  var _34 : usize;
  var _35 : ();
  var _36 : ();
  var _37 : ();
  var _38 : ();
  var _39 : ();
  var _40 : ();
  var _41 : bool;
  var _42 : usize;
  var _43 : usize;
  var _44 : Type.creusotcontracts_std1_vec_vec uint8;
  var _45 : ();
  var _46 : ();
  var _47 : ();
  var _48 : ();
  var _49 : ();
  var _50 : Type.creusotcontracts_std1_vec_vec uint8;
  {
    needle_1 <- needle;
    haystack_2 <- haystack;
    goto BB0
  }
  BB0 {
    i_3 <- (0 : usize);
    goto BB1
  }
  BB1 {
    invariant no_match { forall k : (int) . 0 <= k && k < UInt64.to_int i_3 -> not (MatchAt0.match_at needle_1 haystack_2 k (Seq.length (Model0.model needle_1))) };
    assume { Resolve0.resolve _8 };
    _8 <- i_3;
    _12 <- haystack_2;
    _11 <- Len0.len _12;
    goto BB2
  }
  BB2 {
    _14 <- needle_1;
    _13 <- Len0.len _14;
    goto BB3
  }
  BB3 {
    _10 <- _11 - _13;
    _9 <- _10 + (1 : usize);
    _7 <- _8 < _9;
    switch (_7)
      | False -> goto BB17
      | _ -> goto BB4
      end
  }
  BB4 {
    j_15 <- (0 : usize);
    goto BB5
  }
  BB5 {
    invariant partial_match { MatchAt0.match_at needle_1 haystack_2 (UInt64.to_int i_3) (UInt64.to_int j_15) };
    assume { Resolve0.resolve _19 };
    _19 <- j_15;
    _21 <- needle_1;
    _20 <- Len0.len _21;
    goto BB6
  }
  BB6 {
    _18 <- _19 < _20;
    switch (_18)
      | False -> goto BB12
      | _ -> goto BB7
      end
  }
  BB7 {
    assert { UInt64.to_int j_15 <= Seq.length (Model0.model needle_1) };
    _22 <- ();
    assume { Resolve1.resolve _22 };
    _27 <- needle_1;
    assume { Resolve0.resolve _28 };
    _28 <- j_15;
    _26 <- Index0.index _27 _28;
    goto BB8
  }
  BB8 {
    assume { Resolve2.resolve _25 };
    _25 <- _26;
    assume { Resolve3.resolve _26 };
    _31 <- haystack_2;
    assume { Resolve0.resolve _33 };
    _33 <- i_3;
    assume { Resolve0.resolve _34 };
    _34 <- j_15;
    _32 <- _33 + _34;
    _30 <- Index0.index _31 _32;
    goto BB9
  }
  BB9 {
    assume { Resolve2.resolve _29 };
    _29 <- _30;
    assume { Resolve3.resolve _30 };
    _24 <- _25 <> _29;
    switch (_24)
      | False -> goto BB11
      | _ -> goto BB10
      end
  }
  BB10 {
    _17 <- ();
    assume { Resolve1.resolve _17 };
    goto BB13
  }
  BB11 {
    _23 <- ();
    assume { Resolve1.resolve _23 };
    j_15 <- j_15 + (1 : usize);
    _6 <- ();
    assume { Resolve1.resolve _6 };
    goto BB5
  }
  BB12 {
    _17 <- ();
    assume { Resolve1.resolve _17 };
    goto BB13
  }
  BB13 {
    _16 <- ();
    assume { Resolve1.resolve _16 };
    assert { false };
    _39 <- ();
    assume { Resolve1.resolve _39 };
    assume { Resolve0.resolve _42 };
    _42 <- j_15;
    assume { Resolve0.resolve j_15 };
    _44 <- needle_1;
    _43 <- Len0.len _44;
    goto BB14
  }
  BB14 {
    _41 <- _42 = _43;
    switch (_41)
      | False -> goto BB16
      | _ -> goto BB15
      end
  }
  BB15 {
    assume { Resolve4.resolve needle_1 };
    assume { Resolve4.resolve haystack_2 };
    assume { Resolve0.resolve _0 };
    _0 <- i_3;
    assume { Resolve0.resolve i_3 };
    goto BB19
  }
  BB16 {
    _40 <- ();
    assume { Resolve1.resolve _40 };
    i_3 <- i_3 + (1 : usize);
    _6 <- ();
    assume { Resolve1.resolve _6 };
    goto BB1
  }
  BB17 {
    assume { Resolve4.resolve needle_1 };
    assume { Resolve0.resolve i_3 };
    _5 <- ();
    assume { Resolve1.resolve _5 };
    _4 <- ();
    assume { Resolve1.resolve _4 };
    _50 <- haystack_2;
    assume { Resolve4.resolve haystack_2 };
    _0 <- Len0.len _50;
    goto BB18
  }
  BB18 {
    goto BB19
  }
  BB19 {
    return _0
  }
  
end
